burn-core/src/optim/rmsprop.rs

```diff
@@ -0,0 +1,500 @@
+/// The optimizer can be configured with [RMSPropConfig](RMSPropConfig).
+pub struct RMSProp<B: Backend> {
+    alpha: f32,
+    // epsilon: f32,
+    centered: bool,
+    // momentum: Option<Momentum<B>>,
+    momentum: RMSPropMomentum,
+    weight_decay: Option<WeightDecay<B>>,
+}
+
+/// State of [RMSProp](RMSProp)
+#[derive(Record, Clone, new)]
+pub struct RMSPropState<B: Backend, const D: usize> {
+    square_avg: SquareAvgState<B, D>,
+    centered: CenteredState<B, D>,
+    momentum: RMSPropMomentumState<B, D>,
+}
+
+/// [SquareAvgState](SquareAvgState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct SquareAvgState<B: Backend, const D: usize> {
+    square_avg: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> SquareAvgState<B, D> {
+    /// transform [SquareAvgState] to the next step
+    pub fn transform(alpha: f32, grad: Tensor<B, D>, state: Option<Self>) -> (Tensor<B, D>, Self) {
+        match state {
+            Some(state) => {
+                let square_avg = state
+                    .square_avg
+                    .clone()
+                    .mul_scalar(alpha)
+                    .add(grad.clone().powf(2.).mul_scalar(1. - alpha));
+                (grad, Self { square_avg })
+            }
+            _ => {
+                let square_avg = grad.clone().powf(2.).mul_scalar(1. - alpha);
+                (grad, Self { square_avg })
+            }
+        }
+    }
+
+    fn to_device(mut self, device: &B::Device) -> Self {
+        self.square_avg = self.square_avg.to_device(device);
+        self
+    }
+}
+
+/// [CenteredState](CenteredState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct CenteredState<B: Backend, const D: usize> {
+    grad_avg: Tensor<B, D>,
+    avg: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> CenteredState<B, D> {
+    /// transform [CenteredState] to the next step
+    pub fn transform(
+        alpha: f32,
+        centered: bool,
+        grad: Tensor<B, D>,
+        square_avg_state: SquareAvgState<B, D>,
+        centered_state: Option<Self>,
+    ) -> (Tensor<B, D>, SquareAvgState<B, D>, Self) {
+        if centered {
+            let grad_avg = match centered_state {
+                Some(state) => state
+                    .grad_avg
+                    .clone()
+                    .mul_scalar(alpha)
+                    .add(grad.clone().mul_scalar(1. - alpha)),
+                _ => grad.clone().mul_scalar(1. - alpha),
+            };
+            let avg = square_avg_state
+                .square_avg
+                .clone()
+                .sub(grad_avg.clone().powf(2.));
+            (grad, square_avg_state, Self { grad_avg, avg })
+        } else {
+            let grad_avg = match centered_state {
+                Some(state) => state.grad_avg,
+                _ => Tensor::zeros(grad.shape()),
+            };
+            let avg = square_avg_state.square_avg.clone();
```

Comment:
Having a `centered state` when `centered` is false doesn't seem correct. Maybe we should keep this state as optional 😅.

Reply:
But actually the `avg` is always needed no matter what value the `centered` is, so there should be at least a `centered state`, IMO, to store the `avg`. And that's the question I want to get an advice if we should keep `grad_avg` or `centered state` in `Option` 🤔

Reply:
Oh and the RMSProp's momentum also has this problem. When the `momentum` is less than or equals to 0, the state is useless too. I think this can be wrapped in option 🤔

Reply:
emmm, I'm wondering if it's enough just to wrap these states with Option, because it will occupy the mem space too even it's None with a specefic Tensor type.

Reply:
Buffers are already stored with a pointer, so you don't have to worry, the size of the Tensor struct is minimal.

Reply:
Make it now! I try to only wrap those fields in need.